django/utils/translation/trans_real.py

```diff
@@ -58,6 +58,63 @@ def reset_cache(**kwargs):
         get_supported_language_variant.cache_clear()
 
 
+class TranslationCatalog:
+    """
+    Simulate a dict for DjangoTranslation._catalog so as multiple catalogs
+    with different plural equations are kept separate.
+    """
+    def __init__(self, trans=None):
+        self._catalogs = [trans._catalog.copy()] if trans else [{}]
+        self._plurals = [trans.plural] if trans else [lambda n: int(n != 1)]
+
+    def __getitem__(self, key):
+        for cat in self._catalogs:
+            try:
+                return cat[key]
+            except KeyError:
+                pass
+        raise KeyError
```

Comment:
What would you say for:
```suggestion
        raise KeyError(key)
```

? That would make the error a bit more descriptive.